### PR TITLE
Use cudf::thread_index_type in concatenate.cu.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR uses `cudf::thread_index_type` in `concatenate.cu` to avoid risk of overflow.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by divyegala — APPROVED** [review]

File: cpp/src/copying/concatenate.cu [hunk 1]
```diff
@@ -118,13 +118,15 @@ __global__ void concatenate_masks_kernel(column_device_view const* views,
                                          size_type number_of_mask_bits,
                                          size_type* out_valid_count)
 {
-  size_type mask_index = threadIdx.x + blockIdx.x * blockDim.x;
+  auto tidx = cudf::thread_index_type{threadIdx.x} +
+              cudf::thread_index_type{blockIdx.x} * cudf::thread_index_type{blockDim.x};
```
Comment (ttnghia) [inline cpp/src/copying/concatenate.cu:122]: `cudf::detail::grid_1d::global_thread_id();`.

File: cpp/src/copying/concatenate.cu [hunk 2]
```diff
@@ -141,8 +143,8 @@ __global__ void concatenate_masks_kernel(column_device_view const* views,
       warp_valid_count += __popc(new_word);
     }
 
-    mask_index += blockDim.x * gridDim.x;
-    active_mask = __ballot_sync(active_mask, mask_index < number_of_mask_bits);
+    tidx += cudf::thread_index_type{blockDim.x} * cudf::thread_index_type{gridDim.x};
```
Comment (ttnghia) [inline cpp/src/copying/concatenate.cu:146]: `tidx += stride` where `stride` is extracted.

File: cpp/src/copying/concatenate.cu (hunk 1 above)
Comment (ttnghia) [reply cpp/src/copying/concatenate.cu:122]: Similar for below code.

**Review by PointKernel — APPROVED** [review]

**Review by vuule — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Pull request merged**